### ISSUE TITLE
String and Boolean mutation inside Annotations

### ISSUE THREAD

**Issue by scelestino** [stryker-mutator/stryker4s#95]
### Summary

Strings and Booleans inside Annotations should be escaped like `case`.

```
case class Foo(bar: Option[String],
                       @JsonProperty("type") bazType: Baz)
```

```
annotation argument needs to be a constant; found: scala.sys.`package`.props.get("ACTIVE_MUTATION").orElse[String](scala.sys.`package`.env.get("ACTIVE_MUTATION")) match {
```

### Stryker config

```
stryker4s {
  mutate: ["app/**/*.scala"],
  base-dir: "/home/scelestino/dev/myapp"
}
```

### Stryker4s environment

The SBT branch.

```
stryker4s 0.1.0-SNAPSHOT
```

### Your Environment

Scala 2.12.7
SBT 1.0.3
OS Ubuntu

**Comment by scelestino** [comment]
I think the same. Maybe we can skip the tree if it is an annotation.

I'm going to check that.

**Comment by scelestino** [comment]
There is one point where we can remove this kind of trees from the mutation list.

Currently `MutantFinder.findMutants` uses a standard `collect` that traverse the entire Tree (every node and leaf). As you did with `ImplicitTreeExtensions`, we can create a custom `Traverser` that only visit valid nodes.

Here you are `collect` that doesn't lookup inside Annot nodes (tree for annotations)
```
  val liftedFn = fn.lift
  val buf = scala.collection.mutable.ListBuffer[Unit]()

  object traverser extends Traverser {
    override def apply(tree: Tree): Unit = tree match {
      case Annot(_) => // nothing
      case _ =>
        liftedFn(tree).foreach(buf += _)
        super.apply(tree)
    }
  }
  traverser(exampleTree)
```

Even we can do the same with `Case(pattern, cond, body)` to skip `pattern` (I know that issue is solved, but this is another way to do it)

**Comment by hugo-vrijswijk** [comment]
Funnily enough, I actually wrote a test for an annotation today. Though that was also the first time I did so. 

But don't annotations require constant values? That'd mean we can't put a pattern match in it

**Comment by scelestino** [comment]
> But don't annotations require constant values? That'd mean we can't put a pattern match in it

Exactly, that's why my proposal is skip Annot Tree from MutantFinder.findMutants

**Comment by scelestino** [comment]
@hugo-vrijswijk can you upload that test so I can try to fix it?

**Comment by hugo-vrijswijk** [comment]
That specific test was for work, so I can't upload it (and it was in Java, for that matter). In my experience Scala doesn't use nearly as many annotations as Java, so I think it's completely fine to skip mutating inside annotations. 

As for the custom Traverser, I'm not sure if that's the best way, even though we might have to look at the Traverser anyway because of #93 . I think changing the matching in the `LiteralMutation` would be the way to go since annotations can only contain literals anyway.

**Comment by legopiraat** [comment]
The annotations I've seen so far are hibernate annotations.

**Comment by scelestino** [comment]
I had issues with serialization annotations.

@hugo-vrijswijk are you talking about `StringMutator`? (this issue affect also Booleans)

**Comment by hugo-vrijswijk** [comment]
Fixed by #112
